prov/rxm/src/rxm.h

```diff
@@ -764,34 +761,48 @@ rxm_tx_buf_release(struct rxm_ep *rxm_ep, struct rxm_tx_buf *tx_buf)
 	       (tx_buf->pkt.ctrl_hdr.type == ofi_ctrl_seg_data) ||
 	       (tx_buf->pkt.ctrl_hdr.type == ofi_ctrl_ack));
 	tx_buf->pkt.hdr.flags = 0;
+	rxm_ep->res_fastlock_acquire(&rxm_ep->buf_pools[tx_buf->type].lock);
 	rxm_buf_release(&rxm_ep->buf_pools[tx_buf->type],
 			(struct rxm_buf *)tx_buf);
+	rxm_ep->res_fastlock_release(&rxm_ep->buf_pools[tx_buf->type].lock);
 }
 
 static inline struct rxm_rx_buf *rxm_rx_buf_get(struct rxm_ep *rxm_ep)
 {
-	return (struct rxm_rx_buf *)rxm_buf_get(
+	struct rxm_rx_buf *rx_buf;
+	fastlock_acquire(&rxm_ep->buf_pools[RXM_BUF_POOL_RX].lock);
+	rx_buf = (struct rxm_rx_buf *)rxm_buf_get(
 			&rxm_ep->buf_pools[RXM_BUF_POOL_RX]);
+	fastlock_release(&rxm_ep->buf_pools[RXM_BUF_POOL_RX].lock);
+	return rx_buf;
 }
 
 static inline void
 rxm_rx_buf_release(struct rxm_ep *rxm_ep, struct rxm_rx_buf *rx_buf)
 {
+	fastlock_acquire(&rxm_ep->buf_pools[RXM_BUF_POOL_RX].lock);
 	rxm_buf_release(&rxm_ep->buf_pools[RXM_BUF_POOL_RX],
 			(struct rxm_buf *)rx_buf);
+	fastlock_release(&rxm_ep->buf_pools[RXM_BUF_POOL_RX].lock);
 }
 
 static inline struct rxm_rma_buf *rxm_rma_buf_get(struct rxm_ep *rxm_ep)
 {
-	return (struct rxm_rma_buf *)rxm_buf_get(
+	struct rxm_rma_buf *rma_buf;
+	rxm_ep->res_fastlock_acquire(&rxm_ep->buf_pools[RXM_BUF_POOL_RMA].lock);
+	rma_buf = (struct rxm_rma_buf *)rxm_buf_get(
 			&rxm_ep->buf_pools[RXM_BUF_POOL_RMA]);
+	rxm_ep->res_fastlock_release(&rxm_ep->buf_pools[RXM_BUF_POOL_RMA].lock);
+	return rma_buf;
 }
 
 static inline void
 rxm_rma_buf_release(struct rxm_ep *rxm_ep, struct rxm_rma_buf *rx_buf)
 {
+	rxm_ep->res_fastlock_acquire(&rxm_ep->buf_pools[RXM_BUF_POOL_RMA].lock);
 	rxm_buf_release(&rxm_ep->buf_pools[RXM_BUF_POOL_RMA],
 			(struct rxm_buf *)rx_buf);
+	rxm_ep->res_fastlock_release(&rxm_ep->buf_pools[RXM_BUF_POOL_RMA].lock);
```

Comment:
Did you mean `fastlock_acquire/release`? Ditto for all of them. And why do we need to move the lock / release out of rxm_buf_get / release?

Reply:
He uses the function pointer for the tx pool, but always locks rx.  See rxm_rx_buf_get/release

Reply:
Ah, I see. Perhaps we should have two functions, one that uses locks always and one that uses function pointers. That can keep the repetition to a minimum.

Reply:
> Perhaps we should have two functions, one that uses locks always and one that uses function pointers

Good catch. Done